designer/server/src/main/scala/pl/touk/nussknacker/ui/api/TapirCodecs.scala

```diff
@@ -108,9 +114,30 @@ object TapirCodecs {
     implicit val processNameSchema: Schema[ProcessName] = Schema.string
   }
 
+  object MigrateScenarioRequestV1Codec {
+    // TODO: type me properly, see: https://github.com/TouK/nussknacker/pull/5612#discussion_r1514063218
+    implicit val migrateScenarioRequestV1Schema: Schema[MigrateScenarioRequestV1_14] = Schema.anyObject
+  }
+
+  object MigrateScenarioRequestV2Codec {
+    // TODO: type me properly, see: https://github.com/TouK/nussknacker/pull/5612#discussion_r1514063218
+    implicit val migrateScenarioRequestV2Schema: Schema[MigrateScenarioRequestV1_15] = Schema.anyObject
+  }
+
   object MigrateScenarioRequestCodec {
     // TODO: type me properly, see: https://github.com/TouK/nussknacker/pull/5612#discussion_r1514063218
     implicit val migrateScenarioRequestSchema: Schema[MigrateScenarioRequest] = Schema.anyObject
+
+    implicit val migrateScenarioRequestEncoder: Encoder[MigrateScenarioRequest] = Encoder.instance {
+      case v14: MigrateScenarioRequestV1_14 => v14.asJson
+      case v15: MigrateScenarioRequestV1_15 => v15.asJson
+    }
+
+    implicit val migrateScenarioRequestDecoder: Decoder[MigrateScenarioRequest] = List[Decoder[MigrateScenarioRequest]](
+      Decoder[MigrateScenarioRequestV1_14].widen,
+      Decoder[MigrateScenarioRequestV1_15].widen,
+    ).reduceLeft(_ or _)
```

Comment:
I followed [official](https://circe.github.io/circe/codecs/adt.html#adts-encoding-and-decoding) approach. This has the advantage of not being dependent of some dynamic values.

Reply:
This approach is one of several. the one in which you try to fit it by the structure is highly error-prone and not so straightforward.

IMO it's not request metadata. As I was saying - this is the feature that allows to migrate some version of the scenario to another version of the scenario. It's a domain logic! The version is a domain thing here. I don't think it should be passed with the header.

Reply:
But we are not talking about version of the scenario. We are talking about Nussknacker version which is irrelevant for the targget Nussknacker, since he always lifts up DTO to his highest version.

Reply:
I just realized that we focus on the Nu version. Maybe we should also consider the approach when we focus on the ScenarioData (graph + the other things from the migration DTO) version. Let's say we have V1 now (in Nu 1.14). In Nu 1.15 there are no changes in the ScenarioData, so there is still V1 (so Nu 1.14 and 1.15 works with the V1). In Nu 1.16 there is some change, so we introduce V2. 

Won't it be simpler?

Reply:
Yeah, but OTOH when 1.15 changes nothing then we not introduce anything new. And with current approach we get explicit breaking versions history.